spec/02-integration/04-admin_api/06-certificates_routes_spec.lua

```diff
@@ -156,31 +176,310 @@ describe("Admin API: #" .. kong_config.database, function()
+        json = cjson.decode(body)
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
+      end)
+
+      it("returns a conflict when duplicates snis are present in the request", function()
+
+        local res = assert(client:send {
+          method  = "PUT",
+          path    = "/certificates",
+          body    = {
+            id = cert_bar.id,
+            snis  = "baz.com,baz.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+
+        local body = assert.res_status(409, res)
+        local json = cjson.decode(body)
+
+        assert.equals('duplicate requested sni name baz.com', json.message)
+
+        -- make sure number of snis don't change
+        res = assert(client:send {
+          method  = "GET",
+          path    = "/snis",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
+      end)
+
+      it("returns a conflict when a pre-existing sni present in request is associated to another certificate", function()
+
+        local res = assert(client:send {
+          method  = "PUT",
+          path    = "/certificates",
+          body    = {
+            id = cert_bar.id,
+            snis  = "foo.com,baz.com",
+          },
+          headers = { ["Content-Type"] = "application/x-www-form-urlencoded" },
+        })
+
+        local body = assert.res_status(409, res)
+        local json = cjson.decode(body)
+
+        assert.equals("certificate with id " .. cert_foo.id .. 
+          " in use for entry with name foo.com", json.message)
+
+        -- make sure number of snis don't change
+        res = assert(client:send {
+          method  = "GET",
+          path    = "/snis",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, #json.data)
+        assert.equal(2, json.total)
+
+        -- make sure we did not add any certificate
+        res = assert(client:send {
+          method = "GET",
+          path = "/certificates",
+        })
+
+        body = assert.res_status(200, res)
+        json = cjson.decode(body)
+        assert.equal(2, json.total)
+        assert.equal(2, #json.data)
       end)
```

Comment:
Blocker: we should also add a test case when removing all `snis` from a certificate (ensuring they are removed and ensuring that the resulting attribute is `[]` and not `{}` if we chose to keep this metatable in the `PUT` handler for now).